src/source/CppMarshal.scala

```diff
@@ -113,8 +124,17 @@ class CppMarshal(spec: Spec) extends Marshal(spec) {
 
   def include(ident: String): String = q(spec.cppIncludePrefix + spec.cppFileIdentStyle(ident) + "." + spec.cppHeaderExt)
 
-  private def toCppType(ty: TypeRef, namespace: Option[String] = None): String = toCppType(ty.resolved, namespace)
-  private def toCppType(tm: MExpr, namespace: Option[String]): String = {
+  private def toCppType(ty: TypeRef, scopeNames: Seq[String] = Seq(), namespace: Option[String] = None): String =
+    toCppType(ty.resolved, scopeNames, namespace)
+
+  private def toCppType(tm: MExpr, scopeNames: Seq[String], namespace: Option[String]): String = {
+    def withNamespace(name: String): String = {
+      val ns = namespace match {
+        case Some(ns) => Some(ns)
+        case None => if (scopeNames.contains(name)) Some(spec.cppNamespace) else None
```

Comment:
Yes. I wanted to keep the patch down to a reasonable size, so I tried avoiding changes to the parts of the code that deal with more than just C++. It's indeed possible that the same issue will show up elsewhere, and in fact I just found a bug in this commit (not handling `DEnum` and `DRecord` names as well).

That said, I don't think this change makes it harder to deal with this in the future, and the general tendency of Java/Obj-C code to capitalize type names while method names are lowercase means that the most frequent occurrence of these errors should be covered by this change.

Anyway yeah, I'll add a comment.